pytripgui/canvas_vc/plotter/managers/voi_manager.py

```diff
@@ -0,0 +1,185 @@
+        #     # there is nothing to plot.
+        #     return
+
+        # remove VOI that should not be plotted or updated
+        voi_names_to_remove = [v for v in self._plotted_voi.keys() if v not in [voi.name for voi in vdx.voi_list]]
+        for name in voi_names_to_remove:
+            self._remove_voi_plot(name)
+
+        for voi in vdx.voi_list:
+            logger.debug("plot() voi:{}".format(voi.name))
+
+            current_slice = self._get_current_slice(vdx, voi)
+
+            # remove VOI plot if it does not exist on current slice
+            if current_slice is None:
+                if self._plotted_voi.get(voi.name) is not None:
+                    self._remove_voi_plot(voi.name)
+                continue
+
+            # for a given VOI, the slice viewed may consist of multiple Contours.
+            # contours are in [[x0,y0,z0], [x1,y1,z1], ... [xn,yn,zn]] (mm)
+            # they will be transformed and put into <data>
+            number_of_contours = len(current_slice.contours)
+            for i, _c in enumerate(current_slice.contours):
+                data = np.array(_c.contour) - np.array([vdx.ctx.xoffset, vdx.ctx.yoffset, 0.0])
+                self._plane_points_idx([data], vdx.ctx)  # this transforms the <data> array
+                contour_color = np.array(voi.color) / 255.0
+
+                if _c.contour_closed:
+                    xy = np.concatenate((data, [data[0]]), axis=0)
+                else:
+                    xy = data
+
+                # TODO: what are plotting here? in terms of mm or pixels?
+                if _c.number_of_points() == 1:
+                    # TODO not reworked yet
+                    self._plot_poi(xy[0, 0], xy[0, 1], color=contour_color, legend=voi.name)
+                else:
+                    if self._plotted_voi.get(voi.name) is None:
+                        self._plotted_voi[voi.name] = []
+                    if len(self._plotted_voi[voi.name]) < number_of_contours:
+                        (line, ) = self._axes.plot(xy[:, 0], xy[:, 1], color=contour_color, zorder=15)
+                        self._plotted_voi[voi.name].append(line)
+                        self._blit_manager.add_artist(line)
+                    else:
+                        line = self._plotted_voi[voi.name][i]
+                        line.set_data(xy[:, 0], xy[:, 1])
+
+    def _get_current_slice(self, vdx, voi):
+        _slice = None
+        # TODO create enum class that holds all plane strings
+        if vdx.projection_selector.plane == "Transversal":
+            _slice = voi.get_slice_at_pos(vdx.ctx.slice_to_z(vdx.projection_selector.current_slice_no + 1))
+        elif vdx.projection_selector.plane == "Sagittal":
+            _slice = voi.get_2d_slice(voi.sagittal, vdx.projection_selector.current_slice_no)
+        elif vdx.projection_selector.plane == "Coronal":
+            _slice = voi.get_2d_slice(voi.coronal, vdx.projection_selector.current_slice_no)
+        return _slice

```

Comment:
Add annotation mentioning type of `_slice`

Reply:
done